cpu/asm/asm.go

```diff
@@ -0,0 +1,275 @@
+// Package asm implements Gameboy CPU assembly
+package asm
+
+import (
+	"bytes"
+	"fmt"
+	"strconv"
+	"strings"
+)
+
+// Assembler is a Gameboy Z-80-like assembler.
+type Assembler struct {
+}
+
+// NewAssembler constructs a new [Assembler] object.
+func NewAssembler() *Assembler {
+	return &Assembler{}
+}
+
+func badInstr(i int, instr *Instruction, reason any) error {
+	return fmt.Errorf("bad instruction: line %d: %q: %v", i, instr, reason)
+}
+
+func illegalOperands(i int, instr *Instruction) error {
+	return badInstr(i, instr, "illegal operands")
+}
+
+// Assemble accepts a stream of instructions and assembles them into binary bytes.
+func (a *Assembler) Assemble(instrs ...*Instruction) ([]byte, error) {
+	var buf bytes.Buffer
+
+	for i, instr := range instrs {
+		badInstr := func(reason any) error {
+			return badInstr(i, instr, reason)
+		}
+
+		if err := instr.Err(); err != nil {
+			return nil, badInstr(err)
+		}
+
+		switch instr.Mnemonic {
+		case nop:
+			buf.WriteByte(0x00)
+
+		case ld:
+			if err := a.ld(i, instr, &buf); err != nil {
+				return nil, err
+			}
+
+		default:
+			return nil, badInstr("unknown mnemnonic")
+		}
+	}
+
+	return buf.Bytes(), nil
+}
+
+func Assemble(instrs ...*Instruction) ([]byte, error) {
+	assm := NewAssembler()
+	return assm.Assemble(instrs...)
+}
+
+// Instruction is a in intermediate representation of a Gameboy CPU instruction.
+type Instruction struct {
+	Mnemonic string
+	Bytes    int
+	Cycles   int
+	Operands []Operand
+	err      error
+}
+
+// Err returns the current error assosciated with the Instruction, if it exists.
+func (i *Instruction) Err() error {
+	if i.err != nil {
+		return fmt.Errorf("%s: %v", i.Mnemonic, i.err)
+	}
+	return nil
+}
+
+// String implements fmt.Stringer
+func (i *Instruction) String() string {
+	var opStr []string
+	for _, op := range i.Operands {
+		opStr = append(opStr, op.String())
+	}
+	var builder strings.Builder
+	builder.WriteString(i.Mnemonic)
+	if len(opStr) > 0 {
+		builder.WriteByte(' ')
+		builder.WriteString(strings.Join(opStr, ", "))
+	}
+
+	return builder.String()
+}
+
+// Operand is One of: [Register, Immediate, Pointer]
+type Operand interface {
+	operand()
+	fmt.Stringer
+}
+
+type (
+	Register8   int
+	Reg8        = Register8
+	Register16  int
+	Reg16       = Register16
+	Immediate8  uint8
+	Imm8        = Immediate8
+	Immediate16 uint16
+	Imm16       = Immediate16
+)
+
+// SImm8 is a helper that constructs an Imm8 from a signed 8 bit integer.
+// This is helpful since Go will not automatically convert Signed Int8 to Unsigned Int8 from constants.
+func SImm8(x int8) Imm8 {
+	return signedImm[int8, Imm8](x)
+}
+
+// SImm16 is a helper that constructs an Imm16 from a signed 16 bit integer.
+// This is helpful since Go will not automatically convert Signed Int8 to Unsigned Int8 from constants.
```

Comment:
//  This is helpful since Go will not automatically convert Signed Int16 to Unsigned Int16 from constants.
```suggestion
// This is helpful since Go will not automatically convert Signed Int8 to Unsigned Int8 from constants.
```